File: leetcode/419_battleships_in_a_board.cpp
#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    int countBattleships(vector<vector<char>>& board) {
        int result = 0;
        if (board.empty() || board[0].empty()) {
            return result;
        }

        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] == 'X' &&
                    (i == 0 || board[i - 1][j] != 'X') &&
                    (j == 0 || board[i][j - 1] != 'X')) {
                   result++;
                }
            }
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    vector<vector<char>> board = {
        {'X', '.', '.', 'X'},
        {'.', '.', '.', 'X'},
        {'.', '.', '.', 'X'},
    };

    assert(2 == solution.countBattleships(board));
}

File: leetcode/419_battleships_in_a_board.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class Solution(object):
    def countBattleships(self, board):
        result = 0
        if not board or not board[0]:
            return result

        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                result += 1 if (
                    (i == 0 or board[i - 1][j] != 'X') and
                    (j == 0 or board[i][j - 1] != 'X') and
                    board[i][j] == 'X'
                ) else 0

        return result


if __name__ == '__main__':
    solution = Solution()

    assert 2 == solution.countBattleships([
        ['X', '.', '.', 'X'],
        ['.', '.', '.', 'X'],
        ['.', '.', '.', 'X'],
    ])
